File: src/gr3sync/camera.py
"""HTTP client for the RICOH GR III Wi-Fi API.

The camera runs an unauthenticated HTTP server on ``192.168.0.1:80`` while its
wireless LAN is in AP mode.  Endpoints were recovered from the firmware image by
Dima Kogan (https://notes.secretsauce.net/notes/2022/06/16_ricoh-gr-iiix-80211-reverse-engineering.html)
and the JSON shapes match what ``clyang/GRsync`` has been consuming in practice.

Standard library only, on purpose: the Wi-Fi half of gr3sync must stay
installable with zero dependencies so it can be dropped onto any machine that
happens to be near the camera.
"""

from __future__ import annotations

import contextlib
import json
import shutil
import urllib.error
import urllib.request
from collections.abc import Iterator
from dataclasses import dataclass
from pathlib import Path

from .errors import CameraApiError, HttpError

DEFAULT_HOST = "192.168.0.1"

#: Models whose ``/v1/photos/{dir}/{file}`` download path this client assumes.
KNOWN_MODELS = ("RICOH GR III", "RICOH GR IIIx")

#: The GR II serves image bodies from the bare ``/{dir}/{file}`` path instead.
LEGACY_MODELS = ("RICOH GR II",)


@dataclass(frozen=True)
class PhotoRef:
    """A single file on the camera's card."""

    directory: str
    filename: str

    @property
    def key(self) -> str:
        """Stable identity used for both the local path and the ledger."""
        return f"{self.directory}/{self.filename}"

    @property
    def extension(self) -> str:
        _, _, ext = self.filename.rpartition(".")
        return ext.upper()

    @property
    def is_raw(self) -> bool:
        return self.extension in ("DNG", "RAW")

    @property
    def is_jpeg(self) -> bool:
        return self.extension in ("JPG", "JPEG")

    def __str__(self) -> str:  # pragma: no cover - trivial
        return self.key


@dataclass(frozen=True)
class CameraProps:
    model: str
    firmware: str | None
    serial: str | None
    battery: int | None
    raw: dict

    @property
    def is_legacy_path(self) -> bool:
        return self.model in LEGACY_MODELS


class GRCamera:
    """Talks to the camera over HTTP. Read-only with respect to the SD card."""

    def __init__(
        self,
        host: str = DEFAULT_HOST,
        *,
        timeout: float = 10.0,
        opener: urllib.request.OpenerDirector | None = None,
    ) -> None:
        self.host = host
        self.timeout = timeout
        # A dedicated opener with no proxy handler: the camera AP is a private
        # link, and inheriting an http_proxy from the environment would send
        # every request to a proxy that cannot route to 192.168.0.1.
        self._opener = opener or urllib.request.build_opener(urllib.request.ProxyHandler({}))

    # -- plumbing ---------------------------------------------------------

    def url(self, path: str) -> str:
        return f"http://{self.host}/{path.lstrip('/')}"

    def _open(self, path: str, *, method: str = "GET", body: bytes | None = None, timeout: float | None = None):
        request = urllib.request.Request(self.url(path), data=body, method=method)
        if body is not None:
            request.add_header("Content-Type", "application/json")
        try:
            return self._opener.open(request, timeout=timeout or self.timeout)
        except urllib.error.HTTPError as exc:
            # The camera reports its own error in the JSON envelope. It is not
            # consistent about whether that comes with a matching HTTP status,
            # so an error body is preferred over the status line when present.
            raise self._error_from_body(exc, path) from exc
        except (urllib.error.URLError, OSError) as exc:
            raise HttpError(f"{method} {path} -> {exc}") from exc

    @staticmethod
    def _error_from_body(exc: urllib.error.HTTPError, path: str) -> HttpError:
        try:
            body = json.loads(exc.read())
        except Exception:
            body = None
        if isinstance(body, dict) and body.get("errCode") not in (None, 200):
            return CameraApiError(int(body["errCode"]), str(body.get("errMsg", "")), path)
        return HttpError(f"{path} -> HTTP {exc.code} {exc.reason}")

    def _json(self, path: str, *, method: str = "GET", body: bytes | None = None) -> dict:
        with self._open(path, method=method, body=body) as response:
            payload = response.read()
        try:
            data = json.loads(payload)
        except json.JSONDecodeError as exc:
            raise HttpError(f"{method} {path} -> response was not JSON: {payload[:200]!r}") from exc
        if not isinstance(data, dict):
            raise HttpError(f"{method} {path} -> expected a JSON object, got {type(data).__name__}")
        err_code = data.get("errCode", 200)
        if err_code != 200:
            raise CameraApiError(err_code, str(data.get("errMsg", "")), path)
        return data

    # -- queries ----------------------------------------------------------

    def ping(self) -> bool:
        """Return True when the camera's HTTP server answers."""
        try:
            self._json("/v1/ping")
        except HttpError:
            return False
        return True

    def props(self) -> CameraProps:
        data = self._json("/v1/props")
        battery = data.get("battery")
        return CameraProps(
            model=str(data.get("model", "")),
            firmware=data.get("firmwareVersion") or data.get("version"),
            serial=data.get("serialNo") or data.get("serialNumber"),
            battery=int(battery) if isinstance(battery, (int, float)) else None,
            raw=data,
        )

    def photos(self) -> list[PhotoRef]:
        """List every file on the card, oldest directory first.

        The camera reports ``{"dirs": [{"name": "100RICOH", "files": [...]}]}``
        already in shooting order, and that order is what makes ``--last N``
        meaningful, so it is preserved rather than re-sorted.
        """
        data = self._json("/v1/photos")
        dirs = data.get("dirs")
        if not isinstance(dirs, list):
            raise HttpError(f"/v1/photos -> missing 'dirs' list, got keys {sorted(data)}")
        refs: list[PhotoRef] = []
        for entry in dirs:
            if not isinstance(entry, dict):
                continue
            name = entry.get("name")
            files = entry.get("files")
            if not isinstance(name, str) or not isinstance(files, list):
                continue
            for filename in files:
                if isinstance(filename, str) and filename:
                    refs.append(PhotoRef(directory=name, filename=filename))
        return refs

    def photo_info(self, ref: PhotoRef) -> dict:
        return self._json(f"/v1/photos/{ref.directory}/{ref.filename}/info")

    # -- transfer ---------------------------------------------------------

    def photo_path(self, ref: PhotoRef, *, legacy: bool = False) -> str:
        return f"/{ref.directory}/{ref.filename}" if legacy else f"/v1/photos/{ref.directory}/{ref.filename}"

    def download(
        self,
        ref: PhotoRef,
        destination: Path,
        *,
        legacy: bool = False,
        timeout: float = 120.0,
    ) -> int:
        """Stream one file to ``destination``. Returns the byte count written.

        The body lands in a ``.part`` sibling first and is renamed only after the
        stream completes *and* its length has been checked against the
        advertised ``Content-Length``, so an interrupted sync can never leave a
        truncated JPEG that a later run would mistake for an already-downloaded
        file.  A transfer cut short by the access point dropping is precisely
        the failure this guards, and without the length check it is silent.
        """
        destination.parent.mkdir(parents=True, exist_ok=True)
        partial = destination.with_name(destination.name + ".part")
        try:
            with self._open(self.photo_path(ref, legacy=legacy), timeout=timeout) as response:
                announced = _content_length(response)
                with partial.open("wb") as handle:
                    shutil.copyfileobj(response, handle, length=1 << 18)
            written = partial.stat().st_size
            if written == 0:
                raise HttpError(f"{ref.key} -> downloaded 0 bytes")
            if announced is not None and written != announced:
                raise HttpError(f"{ref.key} -> truncated: got {written} of {announced} bytes")
            partial.replace(destination)
            return written
        except BaseException:
            partial.unlink(missing_ok=True)
            raise

    # -- teardown ---------------------------------------------------------

    def finish_wlan(self) -> None:
        """Ask the camera to drop its access point.

        The AP dies mid-response, so a transport-level failure here is the
        expected outcome and is not surfaced as an error.
        """
        with contextlib.suppress(HttpError):
            self._open("/v1/device/wlan/finish", method="POST", body=b"{}", timeout=3.0).close()

    def finish_device(self) -> None:
        """Ask the camera to power itself off. Same caveat as :meth:`finish_wlan`."""
        with contextlib.suppress(HttpError):
            self._open("/v1/device/finish", method="POST", body=b"{}", timeout=3.0).close()


def _content_length(response) -> int | None:
    """The advertised body length, or None when the camera did not send one."""
    raw = response.headers.get("Content-Length")
    if raw is None:
        return None
    try:
        return int(raw)
    except ValueError:
        return None


def select(
    refs: list[PhotoRef],
    *,
    jpeg: bool = True,
    raw: bool = True,
    last: int | None = None,
    directory: str | None = None,
) -> Iterator[PhotoRef]:
    """Apply the user-facing filters to a photo listing.

    ``last`` counts *selected* files from the end of the listing, so
    ``--last 5 --jpg`` yields five JPEGs rather than five files of which some
    happen to be DNGs.
    """
    chosen = [
        ref
        for ref in refs
        if (directory is None or ref.directory == directory)
        and ((jpeg and ref.is_jpeg) or (raw and ref.is_raw) or (jpeg and raw and not ref.is_jpeg and not ref.is_raw))
    ]
    if last is not None:
        chosen = chosen[-last:] if last > 0 else []
    return iter(chosen)

File: src/camera.rs
//! HTTP client for the RICOH GR III Wi-Fi API.
//!
//! The camera runs an unauthenticated HTTP server on `192.168.0.1:80` while its
//! wireless LAN is in AP mode. The endpoint list was recovered from the
//! firmware image by Dima Kogan (<https://notes.secretsauce.net/notes/2022/06/16_ricoh-gr-iiix-80211-reverse-engineering.html>)
//! and the JSON shapes match what `clyang/GRsync` has been consuming in
//! practice.
//!
//! This client is **read-only with respect to the SD card**. The only state it
//! changes on the camera is bringing the access point and the device down at
//! the end of a sync.

use std::fs::File;
use std::io::Read;
use std::path::{Path, PathBuf};
use std::time::{Duration, Instant};

use serde::Serialize;

use crate::error::{Error, Result};

pub const DEFAULT_HOST: &str = "192.168.0.1";

/// The GR II serves image bodies from the bare `/{dir}/{file}` path instead of
/// `/v1/photos/{dir}/{file}`.
const LEGACY_MODELS: &[&str] = &["RICOH GR II"];

/// A single file on the camera's card.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct PhotoRef {
    pub directory: String,
    pub filename: String,
}

impl PhotoRef {
    pub fn new(directory: impl Into<String>, filename: impl Into<String>) -> Self {
        Self {
            directory: directory.into(),
            filename: filename.into(),
        }
    }

    /// Stable identity, used for both the local path and the ledger.
    pub fn key(&self) -> String {
        format!("{}/{}", self.directory, self.filename)
    }

    pub fn extension(&self) -> String {
        self.filename
            .rsplit_once('.')
            .map(|(_, ext)| ext.to_ascii_uppercase())
            .unwrap_or_default()
    }

    pub fn is_raw(&self) -> bool {
        matches!(self.extension().as_str(), "DNG" | "RAW")
    }

    pub fn is_jpeg(&self) -> bool {
        matches!(self.extension().as_str(), "JPG" | "JPEG")
    }
}

impl std::fmt::Display for PhotoRef {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}/{}", self.directory, self.filename)
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct CameraProps {
    pub model: String,
    pub firmware: Option<String>,
    pub serial: Option<String>,
    pub battery: Option<i64>,
}

impl CameraProps {
    pub fn is_legacy_path(&self) -> bool {
        LEGACY_MODELS.contains(&self.model.as_str())
    }
}

pub struct Camera {
    host: String,
    agent: ureq::Agent,
}

impl Camera {
    pub fn new(host: impl Into<String>, timeout: Duration) -> Self {
        let config = ureq::Agent::config_builder()
            .timeout_global(Some(timeout))
            // Status handling is ours: the camera puts its real error in the
            // JSON envelope and is not consistent about the HTTP status, so a
            // 404 body still has to be read.
            .http_status_as_error(false)
            .build();
        Self {
            host: host.into(),
            agent: config.into(),
        }
    }

    pub fn host(&self) -> &str {
        &self.host
    }

    fn url(&self, path: &str) -> String {
        format!("http://{}/{}", self.host, path.trim_start_matches('/'))
    }

    // -- plumbing ---------------------------------------------------------

    fn get_json(&self, path: &str) -> Result<serde_json::Value> {
        let mut response = self
            .agent
            .get(self.url(path))
            .call()
            .map_err(|e| Error::Http(format!("GET {path} -> {e}")))?;
        let status = response.status().as_u16();
        let body = response
            .body_mut()
            .read_to_vec()
            .map_err(|e| Error::Http(format!("GET {path} -> reading body: {e}")))?;
        Self::parse_envelope(path, status, &body)
    }

    fn parse_envelope(path: &str, status: u16, body: &[u8]) -> Result<serde_json::Value> {
        let value: serde_json::Value = serde_json::from_slice(body).map_err(|_| {
            Error::Http(format!(
                "{path} -> HTTP {status}, response was not JSON: {:?}",
                String::from_utf8_lossy(&body[..body.len().min(200)])
            ))
        })?;
        let code = value.get("errCode").and_then(|c| c.as_i64()).unwrap_or(200);
        if code != 200 {
            return Err(Error::CameraApi {
                code,
                message: value
                    .get("errMsg")
                    .and_then(|m| m.as_str())
                    .unwrap_or_default()
                    .to_string(),
                endpoint: path.to_string(),
            });
        }
        if status != 200 {
            return Err(Error::Http(format!("{path} -> HTTP {status}")));
        }
        Ok(value)
    }

    // -- queries ----------------------------------------------------------

    /// Whether the camera's HTTP server answers. Never errors, so it can be
    /// used directly as a polling predicate.
    pub fn ping(&self) -> bool {
        self.get_json("/v1/ping").is_ok()
    }

    /// Block until [`Camera::ping`] succeeds. Always probes at least once, so a
    /// zero timeout means "check now" rather than "give up immediately".
    pub fn wait_until_up(&self, timeout: Duration) -> Result<()> {
        let deadline = Instant::now() + timeout;
        loop {
            if self.ping() {
                return Ok(());
            }
            if Instant::now() >= deadline {
                return Err(Error::CameraUnreachable {
                    host: self.host.clone(),
                    secs: timeout.as_secs(),
                });
            }
            std::thread::sleep(Duration::from_millis(700));
        }
    }

    pub fn props(&self) -> Result<CameraProps> {
        let value = self.get_json("/v1/props")?;
        let text = |key: &str| {
            value
                .get(key)
                .and_then(|v| v.as_str())
                .map(|s| s.to_string())
        };
        Ok(CameraProps {
            model: text("model").unwrap_or_default(),
            firmware: text("firmwareVersion").or_else(|| text("version")),
            serial: text("serialNo").or_else(|| text("serialNumber")),
            battery: value.get("battery").and_then(|b| b.as_i64()),
        })
    }

    /// List every file on the card, in the order the camera reports it.
    ///
    /// That order is shooting order, and it is what makes `--last N`
    /// meaningful, so it is preserved rather than re-sorted.
    pub fn photos(&self) -> Result<Vec<PhotoRef>> {
        let value = self.get_json("/v1/photos")?;
        // A listing we cannot parse must be an error, never an empty card: the
        // latter would make a sync report success having downloaded nothing.
        let dirs = value
            .get("dirs")
            .and_then(|d| d.as_array())
            .ok_or_else(|| {
                Error::Http(format!(
                    "/v1/photos -> missing 'dirs' list, got keys {:?}",
                    value
                        .as_object()
                        .map(|o| o.keys().cloned().collect::<Vec<_>>())
                        .unwrap_or_default()
                ))
            })?;

        let mut refs = Vec::new();
        for entry in dirs {
            let (Some(name), Some(files)) = (
                entry.get("name").and_then(|n| n.as_str()),
                entry.get("files").and_then(|f| f.as_array()),
            ) else {
                continue;
            };
            for file in files {
                if let Some(filename) = file.as_str().filter(|s| !s.is_empty()) {
                    refs.push(PhotoRef::new(name, filename));
                }
            }
        }
        Ok(refs)
    }

    pub fn photo_info(&self, photo: &PhotoRef) -> Result<serde_json::Value> {
        self.get_json(&format!(
            "/v1/photos/{}/{}/info",
            photo.directory, photo.filename
        ))
    }

    // -- transfer ---------------------------------------------------------

    pub fn photo_path(&self, photo: &PhotoRef, legacy: bool) -> String {
        if legacy {
            format!("/{}/{}", photo.directory, photo.filename)
        } else {
            format!("/v1/photos/{}/{}", photo.directory, photo.filename)
        }
    }

    /// Stream one file to `destination`, returning the byte count written.
    ///
    /// The body lands in a `.part` sibling and is renamed only after the stream
    /// completes *and* its length has been checked against the advertised
    /// `Content-Length`. A transfer cut short by the access point dropping is
    /// exactly what this guards; without the length check it is silent, and the
    /// truncated file would be skipped as "already downloaded" next run.
    pub fn download(&self, photo: &PhotoRef, destination: &Path, legacy: bool) -> Result<u64> {
        let parent = destination.parent().unwrap_or(Path::new("."));
        std::fs::create_dir_all(parent)
            .map_err(|e| Error::io(format!("creating {}", parent.display()), e))?;

        let partial = partial_path(destination);
        let result = self.stream_to(photo, &partial, legacy);
        match result {
            Ok(written) => {
                std::fs::rename(&partial, destination)
                    .map_err(|e| Error::io(format!("renaming {}", partial.display()), e))?;
                Ok(written)
            }
            Err(err) => {
                let _ = std::fs::remove_file(&partial);
                Err(err)
            }
        }
    }

    fn stream_to(&self, photo: &PhotoRef, partial: &Path, legacy: bool) -> Result<u64> {
        let path = self.photo_path(photo, legacy);
        let response = self
            .agent
            .get(self.url(&path))
            .call()
            .map_err(|e| Error::Http(format!("GET {path} -> {e}")))?;

        let status = response.status().as_u16();
        if status != 200 {
            return Err(Error::Http(format!("{} -> HTTP {status}", photo.key())));
        }
        let announced = response
            .headers()
            .get("content-length")
            .and_then(|v| v.to_str().ok())
            .and_then(|v| v.parse::<u64>().ok());

        let mut file = File::create(partial)
            .map_err(|e| Error::io(format!("creating {}", partial.display()), e))?;
        // into_reader(), not read_to_vec(): the latter carries a 10 MB default
        // cap that would silently refuse every DNG off this camera.
        let mut reader = response.into_body().into_reader();
        let written = std::io::copy(&mut reader, &mut file)
            .map_err(|e| Error::io(format!("writing {}", partial.display()), e))?;
        drop(file);

        verify_length(&photo.key(), written, announced)?;
        Ok(written)
    }

    // -- teardown ---------------------------------------------------------

    /// Ask the camera to drop its access point.
    ///
    /// The AP dies mid-response, so a transport-level failure here is the
    /// expected outcome and is deliberately not surfaced.
    pub fn finish_wlan(&self) {
        let _ = self.post_empty("/v1/device/wlan/finish");
    }

    /// Ask the camera to power itself off. Same caveat as [`Camera::finish_wlan`].
    pub fn finish_device(&self) {
        let _ = self.post_empty("/v1/device/finish");
    }

    fn post_empty(&self, path: &str) -> std::result::Result<(), ureq::Error> {
        self.agent
            .post(self.url(path))
            .content_type("application/json")
            .send("{}")?;
        Ok(())
    }
}

/// Defence in depth against a truncated body.
///
/// `ureq` already errors when a length-delimited response is cut short, so in
/// practice this fires only where the transport cannot tell — and it never
/// fires for a close-delimited response, which carries no length to check
/// against. Kept as a pure function so the rule is actually exercised rather
/// than sitting in a branch no test can reach.
fn verify_length(key: &str, written: u64, announced: Option<u64>) -> Result<()> {
    if written == 0 {
        return Err(Error::Http(format!("{key} -> downloaded 0 bytes")));
    }
    match announced {
        Some(expected) if written != expected => Err(Error::Http(format!(
            "{key} -> truncated: got {written} of {expected} bytes"
        ))),
        _ => Ok(()),
    }
}

fn partial_path(destination: &Path) -> PathBuf {
    let mut name = destination.file_name().unwrap_or_default().to_os_string();
    name.push(".part");
    destination.with_file_name(name)
}

/// Read a whole small body. Only used where the payload is known to be tiny.
#[allow(dead_code)]
fn read_all(reader: &mut impl Read) -> std::io::Result<Vec<u8>> {
    let mut buffer = Vec::new();
    reader.read_to_end(&mut buffer)?;
    Ok(buffer)
}

// ---------------------------------------------------------------------------
// Selection
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Copy)]
pub struct Filter<'a> {
    pub jpeg: bool,
    pub raw: bool,
    pub last: Option<usize>,
    pub directory: Option<&'a str>,
}

impl Default for Filter<'_> {
    fn default() -> Self {
        Self {
            jpeg: true,
            raw: true,
            last: None,
            directory: None,
        }
    }
}

/// Apply the user-facing filters to a photo listing.
///
/// `last` counts *selected* files from the end, so `--last 5 --jpg` yields five
/// JPEGs rather than five files of which some happen to be DNGs.
pub fn select(refs: &[PhotoRef], filter: Filter<'_>) -> Vec<PhotoRef> {
    let mut chosen: Vec<PhotoRef> = refs
        .iter()
        .filter(|r| filter.directory.is_none_or(|d| r.directory == d))
        .filter(|r| {
            (filter.jpeg && r.is_jpeg())
                || (filter.raw && r.is_raw())
                // Anything that is neither (a movie, say) only comes along when
                // no format filter was asked for at all.
                || (filter.jpeg && filter.raw && !r.is_jpeg() && !r.is_raw())
        })
        .cloned()
        .collect();
    if let Some(last) = filter.last {
        let skip = chosen.len().saturating_sub(last);
        chosen = chosen.split_off(skip);
    }
    chosen
}

#[cfg(test)]
mod tests {
    use super::*;

    fn refs(keys: &[&str]) -> Vec<PhotoRef> {
        keys.iter()
            .map(|k| {
                let (d, f) = k.split_once('/').unwrap();
                PhotoRef::new(d, f)
            })
            .collect()
    }

    #[test]
    fn photo_ref_classification() {
        assert!(PhotoRef::new("100RICOH", "R0000001.JPG").is_jpeg());
        assert!(PhotoRef::new("100RICOH", "R0000001.jpg").is_jpeg());
        assert!(PhotoRef::new("100RICOH", "R0000001.DNG").is_raw());
        assert!(!PhotoRef::new("100RICOH", "R0000001.MOV").is_jpeg());
        assert_eq!(
            PhotoRef::new("100RICOH", "R0000001.JPG").key(),
            "100RICOH/R0000001.JPG"
        );
    }

    #[test]
    fn download_path_differs_for_the_gr2() {
        let camera = Camera::new("192.168.0.1", Duration::from_secs(1));
        let photo = PhotoRef::new("100RICOH", "R0000001.JPG");
        assert_eq!(
            camera.photo_path(&photo, false),
            "/v1/photos/100RICOH/R0000001.JPG"
        );
        assert_eq!(camera.photo_path(&photo, true), "/100RICOH/R0000001.JPG");
    }

    #[test]
    fn a_short_body_is_refused_when_a_length_was_announced() {
        let err = verify_length("100RICOH/a.DNG", 1024, Some(25_000_000)).unwrap_err();
        assert!(err.to_string().contains("truncated"), "{err}");
        assert!(err.to_string().contains("100RICOH/a.DNG"), "{err}");
    }

    #[test]
    fn an_empty_body_is_refused_even_without_a_length() {
        assert!(verify_length("100RICOH/a.JPG", 0, None).is_err());
    }

    #[test]
    fn a_matching_length_passes_and_a_close_delimited_body_is_accepted() {
        assert!(verify_length("a", 100, Some(100)).is_ok());
        // No Content-Length means nothing to compare against; refusing here
        // would reject every close-delimited response outright.
        assert!(verify_length("a", 100, None).is_ok());
    }

    #[test]
    fn partial_file_sits_beside_the_destination() {
        assert_eq!(
            partial_path(Path::new("/tmp/100RICOH/R0000001.JPG")),
            PathBuf::from("/tmp/100RICOH/R0000001.JPG.part")
        );
    }

    #[test]
    fn an_error_envelope_is_surfaced_even_on_http_200() {
        let body = br#"{"errCode":404,"errMsg":"not found"}"#;
        let err = Camera::parse_envelope("/v1/photos/x/y/info", 200, body).unwrap_err();
        assert!(matches!(err, Error::CameraApi { code: 404, .. }), "{err:?}");
    }

    #[test]
    fn an_error_envelope_is_preferred_over_the_status_line() {
        // The camera is not consistent about pairing its errCode with a
        // matching HTTP status; the body is the authoritative one.
        let body = br#"{"errCode":404,"errMsg":"not found"}"#;
        let err = Camera::parse_envelope("/v1/x", 404, body).unwrap_err();
        assert!(matches!(err, Error::CameraApi { code: 404, .. }), "{err:?}");
    }

    #[test]
    fn a_non_json_body_names_the_endpoint() {
        let err = Camera::parse_envelope("/v1/props", 500, b"<html>oops</html>").unwrap_err();
        assert!(err.to_string().contains("/v1/props"), "{err}");
    }

    #[test]
    fn select_defaults_to_everything() {
        let all = refs(&["100RICOH/a.JPG", "100RICOH/a.DNG", "101RICOH/b.MOV"]);
        assert_eq!(select(&all, Filter::default()).len(), 3);
    }

    #[test]
    fn select_filters_by_format() {
        let all = refs(&["100RICOH/a.JPG", "100RICOH/a.DNG"]);
        let jpeg_only = select(
            &all,
            Filter {
                raw: false,
                ..Default::default()
            },
        );
        assert_eq!(
            jpeg_only
                .iter()
                .map(|r| r.filename.clone())
                .collect::<Vec<_>>(),
            ["a.JPG"]
        );
        let raw_only = select(
            &all,
            Filter {
                jpeg: false,
                ..Default::default()
            },
        );
        assert_eq!(
            raw_only
                .iter()
                .map(|r| r.filename.clone())
                .collect::<Vec<_>>(),
            ["a.DNG"]
        );
    }

    #[test]
    fn last_counts_selected_files_not_listing_entries() {
        // `--last 2 --jpg` must yield two JPEGs, not two of the last four files.
        let all = refs(&[
            "100RICOH/a.JPG",
            "100RICOH/a.DNG",
            "100RICOH/b.JPG",
            "100RICOH/b.DNG",
            "100RICOH/c.JPG",
            "100RICOH/c.DNG",
        ]);
        let picked = select(
            &all,
            Filter {
                raw: false,
                last: Some(2),
                ..Default::default()
            },
        );
        assert_eq!(
            picked
                .iter()
                .map(|r| r.filename.clone())
                .collect::<Vec<_>>(),
            ["b.JPG", "c.JPG"]
        );
    }

    #[test]
    fn last_larger_than_the_card_is_not_an_error() {
        let all = refs(&["100RICOH/a.JPG"]);
        assert_eq!(
            select(
                &all,
                Filter {
                    last: Some(99),
                    ..Default::default()
                }
            )
            .len(),
            1
        );
        assert_eq!(
            select(
                &all,
                Filter {
                    last: Some(0),
                    ..Default::default()
                }
            )
            .len(),
            0
        );
    }

    #[test]
    fn select_filters_by_directory() {
        let all = refs(&["100RICOH/a.JPG", "101RICOH/b.JPG"]);
        let picked = select(
            &all,
            Filter {
                directory: Some("101RICOH"),
                ..Default::default()
            },
        );
        assert_eq!(picked.len(), 1);
        assert_eq!(picked[0].directory, "101RICOH");
    }
}
